<!-- launch file to start map_server, amcl navigation stack, SLAM, move_base and gazebo simulator -->
<launch>
  <include file="$(find turtlebot_navigation)/launch/includes/velocity_smoother.launch.xml"/>
  <include file="$(find turtlebot_navigation)/launch/includes/safety_controller.launch.xml"/> 

  <arg name="map" default="$(find robopatrol)/resources/maps/map.yaml"/>
  <arg name="odom_frame_id"   default="odom"/>
  <arg name="base_frame_id"   default="base_footprint"/>
  <arg name="global_frame_id" default="map"/>
  <arg name="odom_topic" default="odom" />
  <arg name="laser_topic" default="scan" />


  <!-- Run the map server -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map)" output="log">
  </node>

  <!--- Run AMCL (Adaptive Monte Carlo Localisation)
        AMCL is responsible for localizing the robot on the map -->
  <include file="$(find turtlebot_navigation)/launch/includes/amcl.launch.xml" />
  
  <!-- run SLAM -->
  <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="log" required="true">
  </node>

  <!-- run move_base -->
  <node pkg="move_base" type="move_base" name="move_base" output="log" respawn="false" required="true">
    <rosparam file="$(find robopatrol)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find robopatrol)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />   
    <rosparam file="$(find robopatrol)/param/local_costmap_params.yaml" command="load" />   
    <rosparam file="$(find robopatrol)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find robopatrol)/param/dwa_local_planner_params.yaml" command="load" />
    <rosparam file="$(find robopatrol)/param/move_base_params.yaml" command="load" />
    <rosparam file="$(find robopatrol)/param/global_planner_params.yaml" command="load" />
    <rosparam file="$(find robopatrol)/param/navfn_global_planner_params.yaml" command="load" />
    
    <!-- reset frame_id parameters using user input data -->
    <param name="global_costmap/global_frame" value="$(arg global_frame_id)"/>
    <param name="global_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="local_costmap/global_frame" value="$(arg odom_frame_id)"/>
    <param name="local_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="DWAPlannerROS/global_frame_id" value="$(arg odom_frame_id)"/>

    <remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <remap from="scan" to="$(arg laser_topic)"/>
  </node>

  <!-- run the simulation -->
  <include file="$(find turtlebot_gazebo)/launch/turtlebot_world.launch" >
  </include>

  <!-- run autonomous patrol -->
  <node pkg="robopatrol" type="autonomousPatrol" name="Autonomous_RoboPatrol" output="screen" respawn="true">
  </node>
</launch>
